ksql-engine/src/main/java/io/confluent/ksql/topic/TopicDeleteInjector.java

```diff
@@ -0,0 +1,119 @@
+ * {@code DROP [STREAM/TABLE] X DELETE TOPIC}. Any other statements are just
+ * passed through. Furthermore, it will remove the DELETE TOPIC clause from
+ * the statement, indicating that the operation has already been done.
+ *
+ * <p>If the topic being deleted is {@link DataSource.DataSourceSerDe#AVRO},
+ * this injector will also clean up the corresponding schema in the schema
+ * registry.
+ */
+public class TopicDeleteInjector implements Injector {
+
+  private final MetaStore metastore;
+  private final KafkaTopicClient topicClient;
+  private final SchemaRegistryClient schemaRegistryClient;
+
+  public TopicDeleteInjector(final KsqlExecutionContext executionContext) {
+    this(
+        Objects.requireNonNull(executionContext, "executionContext").getMetaStore(),
+        executionContext.getServiceContext().getTopicClient(),
+        executionContext.getServiceContext().getSchemaRegistryClient()
+    );
+  }
+
+  @VisibleForTesting
+  TopicDeleteInjector(
+      final MetaStore metastore,
+      final KafkaTopicClient serviceContext,
+      final SchemaRegistryClient schemaRegistryClient
+  ) {
+    this.metastore = Objects.requireNonNull(metastore, "metastore");
+    this.topicClient = Objects.requireNonNull(serviceContext, "topicClient");
+    this.schemaRegistryClient =
+        Objects.requireNonNull(schemaRegistryClient, "schemaRegistryClient");
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T extends Statement> ConfiguredStatement<T> inject(
+      final ConfiguredStatement<T> statement) {
+    if (!(statement.getStatement() instanceof DropStatement)) {
+      return statement;
+    }
+
+    final DropStatement dropStatement = (DropStatement) statement.getStatement();
+    if (!dropStatement.isDeleteTopic()) {
+      return statement;
+    }
+
+    final String sourceName = dropStatement.getName().getSuffix();
+    final StructuredDataSource<?> source = metastore.getSource(sourceName);
+
+    if (source != null) {
+      try {
+        ExecutorUtil.executeWithRetries(
+            () -> topicClient.deleteTopics(ImmutableList.of(source.getKafkaTopicName())),
+            RetryBehaviour.ALWAYS);
+      } catch (Exception e) {
+        throw new KsqlException("Could not delete the corresponding kafka topic: "
+            + source.getKafkaTopicName(), e);
+      }
+
+      if (source.isSerdeFormat(DataSource.DataSourceSerDe.AVRO)) {
+        try {
+          SchemaRegistryUtil.deleteSubjectWithRetries(
+              schemaRegistryClient,
+              source.getKafkaTopicName() + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX);
+        } catch (final Exception e) {
+          throw new KsqlException("Could not clean up the schema registry for topic: "
+              + source.getKafkaTopicName(), e);
+        }
+      }
+    } else if (dropStatement.getIfExists()) {
+      throw new KsqlException("Could not find source to delete topic for: " + statement);
```

Comment:
I couldn't find what `getIfExists` does. But, does this mean that the metastore might be corrupted and the source is not found? Can this code return the statement without the delete and just log a warning message instead? What do you think?

Reply:
This maintains the same behavior as legacy. The idea is that if I try `DROP STREAM FOO DELETE TOPIC`, but `FOO` doesn't exist, that command should fail.

Reply:
Got it. So there is no `FOO` topic to delete. But is the `DROP` command going to fail if so?

If it is legacy, I think is fine to keep it that way. I just wonder if we should improve this later to allow the `DROP` command to work even if the topic does not exist. It has the same results, doesn't it?

Reply:
I think it makes sense to offer both functionalities (we have `DROP STREAM IF EXISTS FOO` and `DROP STREAM FOO`). The user might want to be alerted that they made a typo if they accidentally type `DROP STREAM FO`, so that they can examine their query and re-issue it later.